File: tests/wasm/test_supertrend.js
/**
 * WASM binding tests for SuperTrend indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    
    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32' 
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);
    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }
    
    testData = loadTestData();
});

test('SuperTrend partial params', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    
    const result = wasm.supertrend(high, low, close, 10, 3.0);
    assert.strictEqual(result.values.length, high.length * 2, 'Result should be 2x input length');
    assert.strictEqual(result.rows, 2, 'Should have 2 rows');
    assert.strictEqual(result.cols, high.length, 'Cols should match input length');
    
    
    const trend = result.values.slice(0, high.length);
    const changed = result.values.slice(high.length);
    
    assert.strictEqual(trend.length, high.length);
    assert.strictEqual(changed.length, high.length);
});

test('SuperTrend accuracy', async () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    const result = wasm.supertrend(high, low, close, 10, 3.0);
    
    
    const trend = result.values.slice(0, high.length);
    const changed = result.values.slice(high.length);
    
    
    const expectedLastFiveTrend = [
        61811.479454208165,
        61721.73150878735,
        61459.10835790861,
        61351.59752211775,
        61033.18776990598,
    ];
    const expectedLastFiveChanged = [0.0, 0.0, 0.0, 0.0, 0.0];
    
    assertArrayClose(
        trend.slice(-5),
        expectedLastFiveTrend,
        1e-4,
        "SuperTrend trend last 5 values mismatch"
    );
    
    assertArrayClose(
        changed.slice(-5),
        expectedLastFiveChanged,
        1e-9,
        "SuperTrend changed last 5 values mismatch"
    );
});

test('SuperTrend zero period', () => {
    
    const high = new Float64Array([10.0, 12.0, 13.0]);
    const low = new Float64Array([9.0, 11.0, 12.5]);
    const close = new Float64Array([9.5, 11.5, 13.0]);
    
    assert.throws(() => {
        wasm.supertrend(high, low, close, 0, 3.0);
    }, /Invalid period/);
});

test('SuperTrend period exceeds length', () => {
    
    const high = new Float64Array([10.0, 12.0, 13.0]);
    const low = new Float64Array([9.0, 11.0, 12.5]);
    const close = new Float64Array([9.5, 11.5, 13.0]);
    
    assert.throws(() => {
        wasm.supertrend(high, low, close, 10, 3.0);
    }, /Invalid period/);
});

test('SuperTrend very small dataset', () => {
    
    const high = new Float64Array([42.0]);
    const low = new Float64Array([40.0]);
    const close = new Float64Array([41.0]);
    
    assert.throws(() => {
        wasm.supertrend(high, low, close, 10, 3.0);
    }, /Invalid period|Not enough valid data/);
});

test('SuperTrend empty input', () => {
    
    const empty = new Float64Array([]);
    
    assert.throws(() => {
        wasm.supertrend(empty, empty, empty, 10, 3.0);
    }, /Empty data/);
});

test('SuperTrend all NaN', () => {
    
    const size = 100;
    const high = new Float64Array(size).fill(NaN);
    const low = new Float64Array(size).fill(NaN);
    const close = new Float64Array(size).fill(NaN);
    
    assert.throws(() => {
        wasm.supertrend(high, low, close, 10, 3.0);
    }, /All values are NaN/);
});

test('SuperTrend NaN handling', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    
    for (let i = 0; i < 5; i++) {
        high[i] = NaN;
        low[i] = NaN;
        close[i] = NaN;
    }
    
    const result = wasm.supertrend(high, low, close, 10, 3.0);
    const trend = result.values.slice(0, high.length);
    const changed = result.values.slice(high.length);
    
    assert.strictEqual(trend.length, high.length);
    assert.strictEqual(changed.length, high.length);
    
    
    for (let i = 0; i < 5; i++) {
        assert(isNaN(trend[i]), `Expected NaN at index ${i}`);
    }
});

test('SuperTrend reinput', () => {
    
    const high = new Float64Array(testData.high);
    const low = new Float64Array(testData.low);
    const close = new Float64Array(testData.close);
    
    
    const result1 = wasm.supertrend(high, low, close, 10, 3.0);
    const trend1 = result1.values.slice(0, high.length);
    
    
    const result2 = wasm.supertrend(trend1, trend1, trend1, 5, 2.0);
    const trend2 = result2.values.slice(0, trend1.length);
    const changed2 = result2.values.slice(trend1.length);
    
    assert.strictEqual(trend2.length, trend1.length);
    assert.strictEqual(changed2.length, trend1.length);
});

test('SuperTrend fast API (no aliasing)', () => {
    const size = 20;
    const high = new Float64Array(size);
    const low = new Float64Array(size);
    const close = new Float64Array(size);
    
    
    for (let i = 0; i < size; i++) {
        close[i] = 50 + i * 0.5 + Math.sin(i) * 2;
        high[i] = close[i] + Math.random() * 2;
        low[i] = close[i] - Math.random() * 2;
    }
    
    
    const highPtr = wasm.supertrend_alloc(size);
    const lowPtr = wasm.supertrend_alloc(size);
    const closePtr = wasm.supertrend_alloc(size);
    const trendPtr = wasm.supertrend_alloc(size);
    const changedPtr = wasm.supertrend_alloc(size);
    
    try {
        
        const memory = new Float64Array(wasm.__wasm.memory.buffer);
        memory.set(high, highPtr / 8);
        memory.set(low, lowPtr / 8);
        memory.set(close, closePtr / 8);
        
        
        wasm.supertrend_into(
            highPtr, lowPtr, closePtr,
            trendPtr, changedPtr,
            size, 5, 2.0
        );
        
        
        const memoryView = new Float64Array(wasm.__wasm.memory.buffer);
        const trend = Array.from(memoryView.slice(trendPtr / 8, trendPtr / 8 + size));
        const changed = Array.from(memoryView.slice(changedPtr / 8, changedPtr / 8 + size));
        
        
        assert.strictEqual(trend.length, size);
        assert.strictEqual(changed.length, size);
        
        
        for (let i = 0; i < changed.length; i++) {
            if (!isNaN(changed[i])) {
                assert(changed[i] === 0.0 || changed[i] === 1.0, 
                    `Changed value at ${i} should be 0 or 1: ${changed[i]}`);
            }
        }
        
        
        const safeResult = wasm.supertrend(high, low, close, 5, 2.0);
        const safeTrend = safeResult.values.slice(0, size);
        const safeChanged = safeResult.values.slice(size);
        
        assertArrayClose(trend, safeTrend, 1e-10, "Fast vs Safe trend mismatch");
        assertArrayClose(changed, safeChanged, 1e-10, "Fast vs Safe changed mismatch");
        
    } finally {
        
        wasm.supertrend_free(highPtr, size);
        wasm.supertrend_free(lowPtr, size);
        wasm.supertrend_free(closePtr, size);
        wasm.supertrend_free(trendPtr, size);
        wasm.supertrend_free(changedPtr, size);
    }
});

test('SuperTrend fast API (with aliasing)', () => {
    const size = 20;
    const data = new Float64Array(size);
    
    
    for (let i = 0; i < size; i++) {
        data[i] = 50 + i * 0.5;
    }
    
    
    const dataPtr = wasm.supertrend_alloc(size);
    const trendPtr = dataPtr; 
    const changedPtr = wasm.supertrend_alloc(size);
    
    try {
        
        const memory = new Float64Array(wasm.__wasm.memory.buffer);
        memory.set(data, dataPtr / 8);
        
        
        wasm.supertrend_into(
            dataPtr, dataPtr, dataPtr,  
            trendPtr, changedPtr,       
            size, 5, 2.0
        );
        
        
        const memoryView = new Float64Array(wasm.__wasm.memory.buffer);
        const trend = Array.from(memoryView.slice(trendPtr / 8, trendPtr / 8 + size));
        const changed = Array.from(memoryView.slice(changedPtr / 8, changedPtr / 8 + size));
        
        
        assert.strictEqual(trend.length, size);
        assert.strictEqual(changed.length, size);
        
        
        let hasValidValues = false;
        for (let i = 5; i < trend.length; i++) {
            if (!isNaN(trend[i])) {
                hasValidValues = true;
                break;
            }
        }
        assert(hasValidValues, 'Should have valid trend values after warmup');
        
    } finally {
        
        wasm.supertrend_free(dataPtr, size);
        wasm.supertrend_free(changedPtr, size);
    }
});

test('SuperTrend batch API', () => {
    const high = new Float64Array(testData.high.slice(0, 100)); 
    const low = new Float64Array(testData.low.slice(0, 100));
    const close = new Float64Array(testData.close.slice(0, 100));
    
    const config = {
        period_range: [8, 12, 2],    
        factor_range: [2.0, 3.0, 0.5] 
    };
    
    const result = wasm.supertrend_batch(high, low, close, config);
    
    
    assert.strictEqual(result.rows, 18, 'Should have 18 rows (2 * 9 combinations)');
    assert.strictEqual(result.cols, 100, 'Should have same length as input');
    assert.strictEqual(result.values.length, 1800, 'Values should have rows*cols values');
    assert.strictEqual(result.periods.length, 9, 'Should have 9 period values');
    assert.strictEqual(result.factors.length, 9, 'Should have 9 factor values');
    
    
    const expectedPeriods = [8, 8, 8, 10, 10, 10, 12, 12, 12];
    const expectedFactors = [2.0, 2.5, 3.0, 2.0, 2.5, 3.0, 2.0, 2.5, 3.0];
    
    assertArrayClose(result.periods, expectedPeriods, 1e-10, "Batch periods mismatch");
    assertArrayClose(result.factors, expectedFactors, 1e-10, "Batch factors mismatch");
    
    
    const singleResult = wasm.supertrend(high, low, close, 10, 2.5);
    const singleTrend = singleResult.values.slice(0, 100);
    
    
    const comboIdx = 4;
    const trendRowIdx = comboIdx * 2;
    const batchTrend = result.values.slice(trendRowIdx * 100, (trendRowIdx + 1) * 100);
    assertArrayClose(batchTrend, singleTrend, 1e-10, "Batch vs single trend mismatch");
});

test('SuperTrend edge cases', () => {
    
    const high = new Float64Array([100.0, 101.0, 102.0, 101.5, 100.5]);
    const low = new Float64Array([99.0, 100.0, 101.0, 100.5, 99.5]);
    const close = new Float64Array([99.5, 100.5, 101.5, 101.0, 100.0]);
    
    const result1 = wasm.supertrend(high, low, close, 2, 0.1);
    const trend1 = result1.values.slice(0, high.length);
    const changed1 = result1.values.slice(high.length);
    
    assert.strictEqual(trend1.length, high.length);
    assert.strictEqual(changed1.length, high.length);
    
    
    const result2 = wasm.supertrend(high, low, close, 2, 10.0);
    const trend2 = result2.values.slice(0, high.length);
    const changed2 = result2.values.slice(high.length);
    
    assert.strictEqual(trend2.length, high.length);
    assert.strictEqual(changed2.length, high.length);
});

File: tests/python/test_supertrend.py
"""
Python binding tests for SuperTrend indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:
    
    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close


class TestSuperTrend:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_supertrend_partial_params(self, test_data):
        """Test SuperTrend with partial parameters - mirrors check_supertrend_partial_params"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        
        trend, changed = ta_indicators.supertrend(high, low, close, 10, 3.0)
        assert len(trend) == len(high)
        assert len(changed) == len(high)
    
    def test_supertrend_accuracy(self, test_data):
        """Test SuperTrend matches expected values from Rust tests - mirrors check_supertrend_accuracy"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        
        trend, changed = ta_indicators.supertrend(high, low, close, 10, 3.0)
        
        assert len(trend) == len(high)
        assert len(changed) == len(high)
        
        
        expected_last_five_trend = [
            61811.479454208165,
            61721.73150878735,
            61459.10835790861,
            61351.59752211775,
            61033.18776990598,
        ]
        expected_last_five_changed = [0.0, 0.0, 0.0, 0.0, 0.0]
        
        
        assert_close(
            trend[-5:], 
            expected_last_five_trend,
            rtol=0.0,
            atol=1e-4,
            msg="SuperTrend trend last 5 values mismatch"
        )
        
        
        assert_close(
            changed[-5:], 
            expected_last_five_changed,
            rtol=0.0,
            atol=1e-9,
            msg="SuperTrend changed last 5 values mismatch"
        )
    
    def test_supertrend_zero_period(self):
        """Test SuperTrend fails with zero period - mirrors check_supertrend_zero_period"""
        high = np.array([10.0, 12.0, 13.0])
        low = np.array([9.0, 11.0, 12.5])
        close = np.array([9.5, 11.5, 13.0])
        
        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.supertrend(high, low, close, 0, 3.0)
    
    def test_supertrend_period_exceeds_length(self):
        """Test SuperTrend fails when period exceeds data length - mirrors check_supertrend_period_exceeds_length"""
        high = np.array([10.0, 12.0, 13.0])
        low = np.array([9.0, 11.0, 12.5])
        close = np.array([9.5, 11.5, 13.0])
        
        
        with pytest.raises(ValueError, match=r"Invalid period|Not enough valid data"):
            ta_indicators.supertrend(high, low, close, 10, 3.0)
    
    def test_supertrend_very_small_dataset(self):
        """Test SuperTrend fails with dataset smaller than period - mirrors check_supertrend_very_small_dataset"""
        high = np.array([42.0])
        low = np.array([40.0])
        close = np.array([41.0])
        
        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.supertrend(high, low, close, 10, 3.0)
    
    def test_supertrend_empty_input(self):
        """Test SuperTrend fails with empty input"""
        empty = np.array([])
        
        
        with pytest.raises(ValueError, match=r"Empty data|All values are NaN"):
            ta_indicators.supertrend(empty, empty, empty, 10, 3.0)
    
    def test_supertrend_all_nan(self):
        """Test SuperTrend with all NaN values"""
        high = np.full(100, np.nan)
        low = np.full(100, np.nan)
        close = np.full(100, np.nan)
        
        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.supertrend(high, low, close, 10, 3.0)
    
    def test_supertrend_nan_handling(self, test_data):
        """Test SuperTrend handles NaN values correctly - mirrors check_supertrend_nan_handling"""
        high = test_data['high'].copy()
        low = test_data['low'].copy()
        close = test_data['close'].copy()
        
        
        high[:5] = np.nan
        low[:5] = np.nan
        close[:5] = np.nan
        
        trend, changed = ta_indicators.supertrend(high, low, close, 10, 3.0)
        assert len(trend) == len(high)
        assert len(changed) == len(high)
        
        
        assert np.all(np.isnan(trend[:5]))
    
    def test_supertrend_reinput(self, test_data):
        """Test SuperTrend applied to its own output - mirrors check_supertrend_reinput"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        
        trend1, changed1 = ta_indicators.supertrend(high, low, close, 10, 3.0)
        
        
        trend2, changed2 = ta_indicators.supertrend(trend1, trend1, trend1, 5, 2.0)
        
        assert len(trend2) == len(trend1)
        assert len(changed2) == len(changed1)
    
    def test_supertrend_batch_single_set(self, test_data):
        """Test batch with single parameter combination"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        
        result = ta_indicators.supertrend_batch(
            high, low, close,
            period_range=(10, 10, 0),
            factor_range=(3.0, 3.0, 0.0)
        )
        
        assert 'trend' in result
        assert 'changed' in result
        assert 'periods' in result
        assert 'factors' in result
        
        
        assert result['trend'].shape == (1, len(high))
        assert result['changed'].shape == (1, len(high))
        
        
        single_trend, single_changed = ta_indicators.supertrend(high, low, close, 10, 3.0)
        assert_close(result['trend'][0], single_trend, rtol=1e-10)
        assert_close(result['changed'][0], single_changed, rtol=1e-10)
    
    def test_supertrend_batch_multiple_params(self, test_data):
        """Test batch with multiple parameter values"""
        high = test_data['high'][:100]  
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        
        
        result = ta_indicators.supertrend_batch(
            high, low, close,
            period_range=(10, 14, 2),     
            factor_range=(2.0, 3.0, 0.5)   
        )
        
        
        assert result['trend'].shape == (9, 100)
        assert result['changed'].shape == (9, 100)
        assert len(result['periods']) == 9
        assert len(result['factors']) == 9
        
        
        expected_periods = [10, 10, 10, 12, 12, 12, 14, 14, 14]
        expected_factors = [2.0, 2.5, 3.0, 2.0, 2.5, 3.0, 2.0, 2.5, 3.0]
        
        assert_close(result['periods'], expected_periods, rtol=1e-10)
        assert_close(result['factors'], expected_factors, rtol=1e-10)
    
    def test_supertrend_streaming(self, test_data):
        """Test SuperTrend streaming functionality"""
        high = test_data['high']
        low = test_data['low']
        close = test_data['close']
        
        
        stream = ta_indicators.SuperTrendStream(period=10, factor=3.0)
        
        
        stream_results = []
        for i in range(min(50, len(high))):
            result = stream.update(high[i], low[i], close[i])
            stream_results.append(result)
        
        
        assert stream_results[0] is None
        
        
        valid_results = [r for r in stream_results if r is not None]
        if valid_results:
            assert isinstance(valid_results[0], tuple)
            assert len(valid_results[0]) == 2
    
    def test_supertrend_kernel_parameter(self, test_data):
        """Test SuperTrend with different kernel parameters"""
        high = test_data['high'][:100]
        low = test_data['low'][:100]
        close = test_data['close'][:100]
        
        
        kernels = [None, 'scalar', 'avx2', 'avx512']
        
        for kernel in kernels:
            try:
                if kernel:
                    trend, changed = ta_indicators.supertrend(high, low, close, 10, 3.0, kernel=kernel)
                else:
                    trend, changed = ta_indicators.supertrend(high, low, close, 10, 3.0)
                
                assert len(trend) == len(high)
                assert len(changed) == len(high)
            except ValueError as e:
                
                if "kernel" not in str(e).lower():
                    raise
    
    def test_supertrend_edge_cases(self):
        """Test SuperTrend with edge cases"""
        
        high = np.array([100.0, 101.0, 102.0, 101.5, 100.5])
        low = np.array([99.0, 100.0, 101.0, 100.5, 99.5])
        close = np.array([99.5, 100.5, 101.5, 101.0, 100.0])
        
        trend, changed = ta_indicators.supertrend(high, low, close, 2, 0.1)
        assert len(trend) == len(high)
        assert len(changed) == len(high)
        
        
        trend, changed = ta_indicators.supertrend(high, low, close, 2, 10.0)
        assert len(trend) == len(high)
        assert len(changed) == len(high)
